algorithms/kernel/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i

```diff
@@ -0,0 +1,398 @@
+
+
+    const Parameter * const parameter = static_cast<const Parameter *>(par);
+    const auto nK = parameter->k;
+
+    const size_t nProbeRows = ntData->getNumberOfRows();
+    const size_t nLabelRows = points->getNumberOfRows();
+    const size_t nDataRows = points->getNumberOfRows() < nLabelRows ? points->getNumberOfRows() : nLabelRows;
+    const size_t nFeatures = points->getNumberOfColumns();
+
+    auto fptype_name = oneapi::internal::getKeyFPType<algorithmFpType>();
+    auto build_options = fptype_name;
+    build_options.add(" -D sortedType=int -D NumParts=16 ");
+
+    services::String cachekey("__daal_algorithms_bf_knn_block_");
+    cachekey.add(fptype_name);
+    {
+        DAAL_ITTNOTIFY_SCOPED_TASK(compute.buildProgram);
+        kernel_factory.build(ExecutionTargetIds::device, cachekey.c_str(), bf_knn_cl_kernels, build_options.c_str());
+    }
+
+    const size_t dataBlockSize = 4096 * 4;
+    const size_t probeBlockSize = (2048 * 4) / sizeof(algorithmFpType);
+    const size_t nParts = 16;
+
+    auto dataSq                 = context.allocate(TypeIds::id<algorithmFpType>(), dataBlockSize,                               &st);
+    auto data_temp              = context.allocate(TypeIds::id<algorithmFpType>(), dataBlockSize * nFeatures,                   &st);
+    auto probe_temp              = context.allocate(TypeIds::id<algorithmFpType>(), probeBlockSize * nFeatures,                   &st);
+    auto distances              = context.allocate(TypeIds::id<algorithmFpType>(), dataBlockSize * probeBlockSize,              &st);
+    auto categories             = context.allocate(TypeIds::id<int>(), probeBlockSize * dataBlockSize,                          &st);
+    auto partialDistances       = context.allocate(TypeIds::id<algorithmFpType>(), probeBlockSize * nK * (nParts + 1),          &st);
+    auto partialCategories      = context.allocate(TypeIds::id<int>(), probeBlockSize * nK * (nParts + 1),                      &st);
+    auto sorted                 = context.allocate(TypeIds::id<int>(), probeBlockSize * nK,                                     &st);
+    auto buffer                 = context.allocate(TypeIds::id<int>(), probeBlockSize * nK,                                     &st);
+    auto radix_buffer           = context.allocate(TypeIds::id<int>(), probeBlockSize * 256,                                    &st);
+    auto classes                = context.allocate(TypeIds::id<int>(), probeBlockSize,                                          &st);
+    auto rndSeq                 = context.allocate(TypeIds::id<algorithmFpType>(), dataBlockSize,                                           &st);
+    auto rndIndices                 = context.allocate(TypeIds::id<size_t>(), dataBlockSize,                                           &st);
+    DAAL_CHECK_STATUS_VAR(st);
+
+    auto rnds = rndSeq.template get<algorithmFpType>().toHost(ReadWriteMode::writeOnly);
```

Comment:
If you are using data on GPU, previously written on CPU, first you should destroy your `rnds` object to release host accessor within it. Please wrap data access on CPU in nested blocks to be sure that data is synchronized after all.